### PR TITLE
MMT-2198 Reading Subscriptions from CMR

### PR DESCRIPTION
**Pull request by ryanmiller-1**
Updated Index controller and view actions to read from CMR
Updated Show/Edit/Update/Destroy to only require a (concept_)id to be passed
Created a before action for Show/Edit/Update/Destroy to find the rest of the information they need from CMR
Updated Edit/Delete buttons where necessary
Updated get_subscriptions
Updated tests - note: the after block deletes could be removed if we could correctly use reset_provider, but it does not work correctly because there is a bug in CMR (ticket written).

### CODE REVIEW

File: app/controllers/subscriptions_controller.rb [hunk 1]
```diff
@@ -68,36 +71,42 @@ def create
...
-    native_id = params.permit(:native_id)['native_id']
 
-    subscription_response = cmr_client.ingest_subscription(@subscription.to_json, current_user.provider_id, native_id, token)
+    subscription_response = cmr_client.ingest_subscription(@subscription.to_json, current_user.provider_id, @native_id, token)
     if subscription_response.success?
       flash[:success] = I18n.t('controllers.subscriptions.update.flash.success')
-      redirect_to subscription_path(subscription_response.parsed_body['result']['concept_id'], subscription: @subscription, native_id: native_id)
+
+      # Hitting CMR too quickly in dev environment.
+      sleep(1) if Rails.env.development? || Rails.env.test?
```
Comment (jceaser) [inline app/controllers/subscriptions_controller.rb:83]: how did this come up? Why did we need to do this?

File: lib/cmr/cmr_client.rb [hunk 2]
```diff
@@ -549,21 +549,13 @@ def check_user_permissions(options, token)
...
-      # so they will change
-      success_response_body =  '{"items":[{"description":"It is a subscription", "user_id":"fake_user_id", "email_address":"fake@fake.fake", "collection_concept_id":"C1200000000-FAKE", "metadata":"thing=stuff&&stuff_id=stuff&&stuff_color=more_stuff"},{"description":"It is a second subscription", "user_id":"fake_user_id", "email_address":"fake2@fake.fake", "collection_concept_id":"C1200000001-FAKE", "metadata":"thing=stuff&&stuff_id=stuff2&&stuff_color=more_stuff2"}]}'
-      Cmr::Response.new(Faraday::Response.new(status: 200, body: JSON.parse(success_response_body)))
-
-      # This error response is provided for viewing and testing a failure
-      # error_response_body = '{"errors":["some error message"]}'
-      # Cmr::Response.new(Faraday::Response.new(status: 400, body: JSON.parse(error_response_body)))
+      url = if Rails.env.development? || Rails.env.test?
+              'http://localhost:3003/subscriptions.umm_json'
+            else
```
Comment (jceaser) [inline lib/cmr/cmr_client.rb:555]: Why the different formats?

Comment (ryanmiller-1) [reply lib/cmr/cmr_client.rb:555]: Error on my part

File: app/controllers/subscriptions_controller.rb [hunk 3]
```diff
@@ -135,4 +144,19 @@ def get_subscriber_email(subscriber_id)
   def add_top_level_breadcrumbs
     add_breadcrumb 'Subscriptions', subscriptions_path
   end
+
+  def fetch_subscription
+    @concept_id = params.permit(:id)['id']
+    options = { 'ConceptId' => @concept_id }
```
Comment (ryanmiller-1) [inline app/controllers/subscriptions_controller.rb:150]: condense

File: app/controllers/subscriptions_controller.rb (hunk 1 above)
Comment (jceaser) [reply app/controllers/subscriptions_controller.rb:83]: I see from the demo why you are doing this, but I think we should find some other solution if possible. If no other solution exists then this idea is not horrible.

File: spec/features/manage_cmr/subscriptions/create_subscriptions_spec.rb [hunk 4]
```diff
@@ -68,6 +68,12 @@
             end
           end
 
+          after do
```
Comment (ryanmiller-1) [inline spec/features/manage_cmr/subscriptions/create_subscriptions_spec.rb:71]: add comment for all these afters that they should be superfluous when we can reset_provider

File: lib/cmr/cmr_client.rb (hunk 2 above)
Comment (ryanmiller-1) [reply lib/cmr/cmr_client.rb:555]: addressed

File: app/controllers/subscriptions_controller.rb (hunk 1 above)
Comment (ryanmiller-1) [reply app/controllers/subscriptions_controller.rb]: addressed, will comment to explain below

**Comment by ryanmiller-1** [comment]
I removed the waits in development/test by adding a 'get_latest_revision' function.  It will try 20 times to get the revision specified or later.  This mimics what we do for publishing collections.  If it fails to find it, it will log and display an error warning (parallel to the one we show with collections).

File: app/controllers/subscriptions_controller.rb [hunk 5]
```diff
@@ -135,4 +133,22 @@ def get_subscriber_email(subscriber_id)
...
     add_breadcrumb 'Subscriptions', subscriptions_path
   end
+
+  def fetch_subscription
+    permitted_params = params.permit(:id, :revision_id)
+    @concept_id = permitted_params['id']
+
+    subscription = get_latest_revision(@concept_id, permitted_params['revision_id'] || 0)
+
+    if subscription
```
Comment (charleshuang80) [inline app/controllers/subscriptions_controller.rb:143]: i would suggest using `fetch` in here as well

File: app/views/subscriptions/index.html.erb [hunk 6]
```diff
@@ -10,10 +10,15 @@
       <p><%= page_entries_info @subscriptions, entry_name: 'Subscription' %></p>
     <% end %>
 
+    <p>
+      <em><span class="fa fa-info-circle"></span> Subscription operations may take some time. If you are not seeing what you expect below, please <%= link_to 'refresh the page', 'javascript:history.go(0)', class: 'eui-btn--link link-lift' %>.</em>
```
Comment (charleshuang80) [inline app/views/subscriptions/index.html.erb:14]: if we are doing the 'try 20 times', do we still need this?

File: app/views/subscriptions/index.html.erb [hunk 7]
```diff
@@ -27,26 +32,33 @@
...
             <tr>
-              <td>
-                <%= subscription['description'] %>
+              <td class='wrapping-td'>
+                <div class='wrapping-div'>
+                  <%= link_to subscription['Name'], subscription_path(subscription['ConceptId']) %>
+                </div>
               </td>
+              <td><%= subscription['CollectionConceptId'] %></td>
               <!-- TODO: need to put a limit or char cutoff for query? -->
```
Comment (charleshuang80) [inline app/views/subscriptions/index.html.erb:41]: still need comment?

File: spec/features/manage_cmr/subscriptions/index_subscriptions_spec.rb [hunk 8]
```diff
@@ -17,54 +30,73 @@
...
+    context 'when viewing the index page' do
+      before do
+        visit subscriptions_path
+      end
+
+      # These tests can be improved when we can reset_provider. They should have
+      # only 2 subscriptions on them, but if the subscription tests are run
+      # together, CMR does not always finish deleting subscriptions from other
+      # tests before starting this one, so it fails at the commented out line.
+      it 'displays sample subscriptions' do
```
Comment (charleshuang80) [inline spec/features/manage_cmr/subscriptions/index_subscriptions_spec.rb:44]: 1. I don't know if 'sample' is the right descriptor
2. I think you want to be more explicit here that it is showing the subscriptions but not the links for which they don't have access

File: app/concerns/cmr_subscriptions.rb [hunk 9]
```diff
@@ -0,0 +1,26 @@
+module CMRSubscriptions
+  extend ActiveSupport::Concern
+
+  def get_latest_revision(concept_id, revision_id)
+    attempts = 0
```
[the remark this answers is not in the document]
Comment (ryanmiller-1) [reply app/concerns/cmr_subscriptions.rb:8]: done

File: app/controllers/subscriptions_controller.rb (hunk 5 above)
Comment (ryanmiller-1) [reply app/controllers/subscriptions_controller.rb:145]: done

File: app/views/subscriptions/index.html.erb (hunk 6 above)
Comment (ryanmiller-1) [reply app/views/subscriptions/index.html.erb]: Spoke with PO, she wanted a new message in its place.

**Pull request merged**